File: dichotomy/dicho.m
function [x,n] = dicho(f, a, b, e)

    % we will be using annonymos function instead of inline
    % since it's going to be deprecated soon

    fx = @(x) eval(f);
    
    n = 0;
    
    while abs(b - a) > e
        
        n = n + 1;
        
        x = (a + b) / 2;
        
        if fx(a) * fx(x) < 0
            b = x;
        else
            a = x;
        end
        
    end
    
    fprintf("solution: %f \niterations counts: %d", x, n);
    
end

File: dichotomy/dicho.rs
fn dicho(f: &dyn Fn(f64) -> f64, a: f64, b: f64, e: f64) -> (f64, u32) {
    let mut n = 0;
    let mut x = 0.0;

    while (b - a).abs() > e {
        n += 1;
        x = (a + b) / 2.0;

        if f(a) * f(x) < 0.0 {
            b = x;
        } else {
            a = x;
        }
    }

    (x, n)
}


